### PR TITLE
mantid-developer metapackage

### LINKED ISSUE #37627: Consolidate dependency requirements
**Describe the outcome that is desired.**
Dependencies and version pinning requirements are currently spread across several files: the three `mantid-developer` environment files and the conda-build recipes. This duplication can easily lead to version pinnings becoming out of sync between developer environments and the conda-build recipes, meaning that we are testing against versions of dependencies that are different to the ones we are shipping against.

There are two separate but related outcomes that would be nice to resolve by consolidating requirements:
- Specify our dependency versions in one place to be shared between the developer and conda-build environments.
- Have a single mantid-developer environment definition to be shared between all OS. Currently there is no support for specifying that a particular package is only installed on a single OS (see [here](https://github.com/conda/conda/issues/8089) for a discussion on that).

**Describe any solutions you are considering**
1.  Create a `mantid-developer` [conda metapackage](https://docs.conda.io/projects/conda-build/en/latest/resources/commands/conda-metapackage.html) to be uploaded to the `mantid` channel.
- Advantages:
   - can easily use the same pinnings as the other package recipes from `conda_build_config.yaml`
   - developers can easily create multiple development environments for testing changes to dependencies without having to rename the environment name in the current mantid-developer file.
- Disadvantages:
  - would require an additional stage in the nightly pipeline, and at the start of all PR checks to re-build the mantid-developer conda package. In the nightly pipeline, we would deploy the package. In PR checks we would first check whether there are any changes to the recipe, and then build and install a local metapackage.
  - new metapackage would only be available after it has been deployed by the nightly pipeline, and wouldn't be directly in sync with `main`
  - might cause issues if there are differences between developer environments between `main` and `release-next`

2.  Write a python script to parse a single yaml file cantaining OS selectors to generate an environment file and inject the version constraints from `conda_build_config.yaml`. This is the approach that was taken by the mantid imaging team (see [here](https://github.com/mantidproject/mantidimaging/blob/1cd9d21191eef1df5ff0deec0b0a0db4272397f4/setup.py#L83)), where you run this command to create the developer environment:
`python3 ./setup.py create_dev_env`
- Advantages:
   - developer env is always configured **directly** from source
   - does not require potentially lengthy conda-build step
   - no need to host a metapackage on the mantid channel
- Disadvantages:
   - it's not a standard way of creating a conda environment
   - may be introducing new bugs

**Additional context**
This is essentially the same as this issue: https://github.com/mantidproject/mantid/issues/33442, but since the conda recipes have moved into the mantid repo, a lot of the information there is out of date.

### PR DESCRIPTION
**Pull request by thomashampson**
### Description of work
Adds a new Conda recipe that generates a `mantid-developer` Conda [metapackage](https://docs.conda.io/projects/conda/en/latest/user-guide/concepts/packages.html#metapackages). The metapackage is a replacement for the `mantid-developer-{linux,osx,win}.yml` environment files that were used to create development environments. The major advantage of this is that dependency version numbers only appear in one place. There were still a few instances where the version pinning was different in the developer and packaging environments. Those should all be resolved now.

#### Summary of work
- Consolidate all developer environments into a single metapackage recipe, and share the version pinnings with the other conda recipes.
- Configure the build scripts so that the metapackage is built locally from source and installed from the local channel before every build. Previously the mantid-developer environment was created from the environment files before each build.

Fixes #37627 

#### Further detail of work
- PR builds will build the metapackage before each job run and install the locally built metapackage. This ensures that any changes to the recipe are tested before they are merged.
- A new metapackage will be uploaded along with the other Conda packages to our Conda channel.

### To test:
- All stages in this pipeline build should succeed, and it should publish the mantid-developer metapackage to the `metapackage_test` label:
https://builds.mantidproject.org/job/build_packages_from_branch/880/
- Try to install the metapackage by:
```
mamba create -n mantid-developer-test
mamba activate mantid-developer-test
mamba install -c mantid/label/metapackage_test mantid-developer
```
- Check that the developer documentation changes are good

---

### Reviewer

Please comment on the points listed below ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)).
**Your comments will be used as part of the gatekeeper process, so please comment clearly on what you have checked during your review.** If changes are made to the PR during the review process then your final comment will be the most important for gatekeepers. In this comment you should make it clear why any earlier review is still valid, or confirm that all requested changes have been addressed.

#### Code Review

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?
- Do the release notes conform to the [release notes guide](https://developer.mantidproject.org/Standards/ReleaseNotesGuide.html)?

#### Functional Tests

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### Gatekeeper

If you need to request changes to a PR then please add a comment and set the review status to "Request changes". This will stop the PR from showing up in the list for other gatekeepers.

### CODE REVIEW

File: conda/recipes/mantid-developer/meta.yaml [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+    - lib3mf # [win]
+    - libboost-devel {{ libboost_devel }}
+    - libboost-python-devel {{ libboost_python_devel }}
+    - libopenblas {{ libopenblas }} # [osx or linux]
+    - librdkafka {{ librdkafka }}
+    - matplotlib {{ matplotlib }}
+    - muparser {{ muparser }}
+    - nexus {{ nexus }}
+    - ninja {{ ninja }}
+    - numpy>=1.24,<1.27 # This is intentionally different to conda_build_config.yaml, will address later.
```
Comment (robertapplin) [inline conda/recipes/mantid-developer/meta.yaml:30]: Out of interest, why is this different?

File: conda/recipes/mantid-developer/meta.yaml [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+    - python {{ python }}
+    - python.app # [osx]
+    - pyyaml {{ pyyaml }}
+    - qscintilla2 {{ qscintilla2 }}
+    - qt {{ qt_main }}
+    - qtconsole {{ qtconsole }}
+    - qtpy {{ qtpy }}
+    - qt-gtk-platformtheme # [linux]
+    - quasielasticbayes
+    - requests>=2.25.1
```
Comment (robertapplin) [inline conda/recipes/mantid-developer/meta.yaml:48]: And is this version also intentionally not put in the conda_build_config.yaml?

**Review by robertapplin — COMMENTED** [review]
Having the version numbers of all our dependencies in one location is a big improvement and will help ensure consistency between our testing and deployment packages.

The code changes look good, I have grepped for instances of `mantid-developer-<OS>` and didn't find anything that needs changing. I had one or two questions just for my own understanding.

I will follow the testing instructions next

File: conda/recipes/mantid-developer/meta.yaml (hunk 2 above)
Comment (robertapplin) [reply conda/recipes/mantid-developer/meta.yaml:48]: From Tom: Requests is only found in the developer environment and not in any of the package environments. In future might be worth checking why this is (but out of scope for this PR)

File: conda/recipes/mantid-developer/meta.yaml (hunk 1 above)
Comment (robertapplin) [reply conda/recipes/mantid-developer/meta.yaml:30]: From Tom:
- We allow numpy package 1.24,1.25,1.26 in our developer environments
- We build all our packages against 1.24

**Review by robertapplin — COMMENTED** [review]
The code changes look good and make sense. I have built mantid locally on Windows using Ninja within a new mantid-developer environment, and it works fine. Also built AllTests and ran the system tests

Is it worth me building in visual studio to check its working from there?

**Comment by jclarkeSTFC** [comment]
> The code changes look good and make sense. I have built mantid locally on Windows using Ninja within a new mantid-developer environment, and it works fine. Also built AllTests and ran the system tests
> 
> Is it worth me building in visual studio to check its working from there?

I built the solution with Visual Studio this morning and it worked fine

**Review by robertapplin — APPROVED** [review]
Happy to approve, it seems to be working well. It builds fine on all operation systems in the build package from branch pipeline, and it also builds Ok with VS on Windows. Code changes look good and my comments have been addressed

**Pull request merged**